sdcm/utils/libssh2_client/__init__.py

```diff
@@ -0,0 +1,750 @@
+        else:
+            end_time = float_info.max
+        while reader.is_alive() or reader.stdout.qsize() or reader.stderr.qsize():
+            if time() > end_time:
+                reader.stop()
+                return False
+            if stdout_stream is not None:
+                try:
+                    stdout = reader.stdout.get(timeout=timeout_read_data_chunk)
+                    data = stdout.decode(encoding) + '\n'
+                    stdout_stream.write(data)
+                    for watcher in watchers:
+                        watcher.submit_line(data)
+                except:  # pylint: disable=bare-except
+                    pass
+            if stderr_stream is not None:
+                if reader.stderr.qsize():
+                    try:
+                        stderr = reader.stderr.get(timeout=timeout_read_data_chunk)
+                        data = stderr.decode(encoding) + '\n'
+                        stderr_stream.write(data)
+                        for watcher in watchers:
+                            watcher.submit_line(data)
+                    except:  # pylint: disable=bare-except
+                        pass
+        return True
+
+    @staticmethod
+    def _process_output_no_watchers(  # pylint: disable=too-many-arguments
+            session: Session, channel: Channel, encoding: str, stdout_stream: StringIO,
+            stderr_stream: StringIO, timeout: NullableTiming, timeout_read_data_chunk: NullableTiming) -> bool:
+        eof_result = stdout_size = stderr_size = LIBSSH2_ERROR_EAGAIN
+        if timeout:
+            end_time = time() + timeout
+        else:
+            end_time = float_info.max
+        while eof_result == LIBSSH2_ERROR_EAGAIN or stdout_size == LIBSSH2_ERROR_EAGAIN or stdout_size > 0 or \
+                stderr_size == LIBSSH2_ERROR_EAGAIN or stderr_size > 0:  # pylint: disable=consider-using-in
+            if time() > end_time:
+                return False
+            with session.lock:
+                if stdout_size == LIBSSH2_ERROR_EAGAIN and stderr_size == LIBSSH2_ERROR_EAGAIN:  # pylint: disable=consider-using-in
+                    session.simple_select(timeout=timeout_read_data_chunk)
+                eof_result = channel.wait_eof()
+                stdout_size, stdout_chunk = channel.read()
+                stderr_size, stderr_chunk = channel.read_stderr()
+            if stdout_chunk and stdout_stream is not None:
+                stdout_stream.write(stdout_chunk.decode(encoding))
+            if stderr_chunk and stderr_stream is not None:
+                stderr_stream.write(stderr_chunk.decode(encoding))
+        return True
+
+    def check_if_alive(self, timeout: NullableTiming = __DEFAULT__):
+        """Check and return if endpoint is capable of running commands
+        """
+        if timeout is __DEFAULT__:
+            timeout = self.timings.check_if_alive_timeout
+        return self.run('true', timeout=timeout).ok
```

Comment:
This is pretty heavy way to ensure that connection is alive. Is there any other way to do this?

Reply:
It depends on definition of alive, you can use keepalive_send or open_session, but even if both succeded it does not mean that whole thing will work.

Most closest thing is to make open_session.

Reply:
OK, let's have it as is, but need to ensure that we don't call it too often.